### PR TITLE
llvm-to-smt: Support llvm.*.with.overflow.* intrinsics and extractvalue

### PR DESCRIPTION
**Pull request by harishankarv**
This commits support for llvm add/sub intrinsics with overflow detection: llvm.sadd.with.overflow.*, llvm.uadd.with.overflow.*, llvm.ssub.with.overflow.*, llvm.usub.with.overflow.* . It also adds limited support for extractvalue instructions to support the intrinsics. This is needed in order to support BPF_ADD and BPF_SUB instructions in v6.10 [1].

The return type of the add with overflow intrinsic is an aggregate type. The types are automatically inferred and appropriate sized bitvectors are used (i32/i64/any other size) in the formula. This commit also adds support for extractvalue instructions, that typically occur *after* the intrinsic call instructions.

See the commit message and comments for details. 

[1]: https://github.com/bpfverif/agni/issues/44

### CODE REVIEW

**Review by pchaigno — COMMENTED** [review]
FYI, the same issue affects "sub with overflow". It won't reproduce with the upstream commit I indicated in https://github.com/bpfverif/agni/issues/44, but with the commit right after, [`deac5871eb07`](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=deac5871eb0751454cb80b3ff6b69e42a6c1bab2). So using `deac5871eb07` for reproduction should show the issues for both BPF_ADD and BPF_SUB.

File: llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp
Comment (harishankarv) [reply llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp:2150]: You are right, that's a typo!

File: llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp
Comment (harishankarv) [reply llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp:2156]: Thanks for the catch. There should be a statement.
```
BBAsstVecIter->second.push_back(resultExprAdd);
```
But yes, still working on this and that's why kept it in draft mode :D

Comment (pchaigno) [reply llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp:2156]: :innocent:

**Comment by harishankarv** [comment]
> FYI, the same issue affects "sub with overflow". It won't reproduce with the upstream commit I indicated in #44, but with the commit right after, [`deac5871eb07`](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=deac5871eb0751454cb80b3ff6b69e42a6c1bab2). So using `deac5871eb07` for reproduction should show the issues for both BPF_ADD and BPF_SUB.

Thanks for the exact commit!  I'll address this in a separate commit.

**Comment by harishankarv** [comment]
Z3 distinguishes between overflows and underflows:  `BVAddNoOverflow,  BVAddNoUnderflow` and `BVAddSubOverflow,  BVAddSubUnderflow`. 

Considering only addition, it seems that per Z3, an overflow occurs when there is a positive overflow (sum of two positive numbers yields a negative result), and an underflow occurs when there is a negative overflow (sum of two negative numbers yields a positive result). I checked this in Z3py. 

Now from the reading of the semantics of [`llvm.sadd.with.overflow` and `llvm.uadd.with.overflow`](https://llvm.org/docs/LangRef.html#arithmetic-with-overflow-intrinsics), it seems that this llvm instrinsic is meant to detect *both* positive and negative overflows.

 It seems we will have to use both `BVAddNoOverflow` and `BVAddNoUnderflow` in conjunction to detect "no" overflow: `And(BVAddNoOverflow, BVAddNoUnderflow) ==> overflow didn't happen`

**Comment by harishankarv** [comment]
Positive overflow (detected only by `BVAddNoOverflow`)
```
   01100100   (100)
+  00110010   (50)
-----------
   10010110   (-106)
```
Negative overflow (detected only by `BVAddNoUnderflow`)
```
   10111010   (-70)
+  11000100   (-60)
-----------
   01111110   (+126)
```

File: llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp
Comment (pchaigno) [inline llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp:2187]: Could we do a `switch (intrinsicInstID)` instead of `if/else` and error in the `default` case (instead of having to do it in the caller)?

File: llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp
Comment (pchaigno) [inline llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp:2166]: Regarding Z3's semantics for underflows: Yeah, very weird (and annoying for us). I've never seen underflows being defined like that :confused:

**Review by pchaigno — APPROVED** [review]
Nice, nice, nice! :smiley: 

I think you can also add the four instructions to https://github.com/bpfverif/agni/blob/24276bf6f6cdb96e974fd8667274280468453cca/.github/workflows/end-to-end.yaml#L36-L37 so that we'll have coverage in the scheduled workflow as soon as this is merged.

File: llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp
Comment (harishankarv) [reply llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp:2187]: It's a good idea to use switch statements. 

I thought it makes sense to keep the check in the caller, because this particular function `handleAddSubWithOverflowIntrinsic` is meant to handle only the four intrinsics add/sub/overflow/underflow.  So any function that calls it should make sure it is one of those.

**Pull request merged**
